integr8ly/cloud-resource-operator
# MGDAPI-4992 implement gcp redis update/upgrade logic

## Overview

Jira: https://issues.redhat.com/browse/MGDAPI-4992

## Verification

- Provision a GCP cluster
  - From the master branch of the [Delorean](https://github.com/integr8ly/delorean) repo create a file in the root of the repo called gcp_service_account.json. Request GCP credentials and add these to the file
  - Run:
    ```
    make -f make/ocm.mk ocm/cluster.json BYOC=true CLOUD_PROVIDER=gcp OCM_CLUSTER_REGION=europe-west2
    ```
  - Edit the cluster.json file to update the name (and remove the expiration timestamp if necessary). You can also reduce the number of nodes to 2
  - Run:
    ```
    make ocm/cluster/create
    ```
- Create a Redis instance
  - Checkout my pull request
    ```
    gh pr checkout 590
    ```
  - Log in to your cluster and run:
    ```
    PROVIDER=gcp make cluster/prepare
    ```
  - Run CRO:
    ```
    make run
    ```
  - Create a Redis instance (v5.0):
    ```
    oc patch configmap cloud-resources-gcp-strategies -n cloud-resource-operator --type=json -p='[{"op":"replace","path":"/data/redis","value":"{\"development\":{\"region\":\"\",\"projectID\":\"\",\"createStrategy\":{\"instance\":{\"redis_version\":\"REDIS_5_0\"}},\"deleteStrategy\":{}}}"}]'
    PROVIDER=gcp make cluster/seed/redis
    ```
  - Wait until it provisions and check that the instance has been created on GCP:
    ```
    gcloud redis instances list --region=europe-west2
    gcloud redis instances describe <INSTANCE_NAME> --region=europe-west2
    ```
  - Patch the strategy config map which will trigger an update on the previously created Redis instance:
    ```
    oc patch configmap cloud-resources-gcp-strategies -n cloud-resource-operator --type=json -p='[{"op":"replace","path":"/data/redis","value":"{\"development\":{\"region\":\"\",\"projectID\":\"\",\"createStrategy\":{\"instance\":{\"redis_version\":\"REDIS_5_0\",\"memory_size_gb\":2,\"labels\":{\"test-label-key\":\"test-label-value-new\"},\"redis_configs\":{\"activedefrag\":\"yes\"}}},\"deleteStrategy\":{}}}"}]'
    oc patch redis example-redis -n cloud-resource-operator --type=json -p='[{"op":"replace","path":"/spec/maintenanceWindow","value":true}]'
    ```
  - Wait until it updates and check that the instance has all updates listed in the config map applied:
    ```
    gcloud redis instances list --region=europe-west2
    gcloud redis instances describe <INSTANCE_NAME> --region=europe-west2
    ```
  - Patch the strategy config map which will trigger an upgrade on the previously created Redis instance:
    ```
    oc patch configmap cloud-resources-gcp-strategies -n cloud-resource-operator --type=json -p='[{"op":"replace","path":/data/redis,"value":"{\"development\":{\"region\":\"\",\"projectID\":\"\",\"createStrategy\":{\"instance\":{\"redis_version\":\"REDIS_6_X\"}},\"deleteStrategy\":{}}}"}]'
    oc patch redis example-redis -n cloud-resource-operator --type=json -p='[{"op":"replace","path":"/spec/maintenanceWindow","value":true}]'
    ```
  - Wait until it upgrades and check that the instance has the upgraded version listed in the config map applied:
    ```
    gcloud redis instances list --region=europe-west2
    gcloud redis instances describe <INSTANCE_NAME> --region=europe-west2
    ```
  - Finally, delete the Redis instance and wait for it to complete:
    ```
    oc delete redis example-redis -n cloud-resource-operator
    ```
